python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,927 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ["Database", "TransactionInterruption", "ReadOnlyDatabaseError", "StaticTablesContext",
+           "SynchronizationConflict"]
+
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+)
+
+import sqlalchemy
+
+from .. import ddl
+
+
+def _checkExistingTableDefinition(name: str, spec: ddl.TableSpec, inspection: Dict[str, Any]):
+    """Test that the definition of a table in a `ddl.TableSpec` and from
+    database introspection are the consistent.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of the table (only used in error messages).
+    spec : `ddl.TableSpec`
+        Specification of the table.
+    inspection : `dict`
+        Dictionary returned by
+        `sqlalchemy.engine.reflection.Inspector.get_columns`.
+
+    Raises
+    ------
+    RuntimeError
+        Raised if the definitions are inconsistent.
+    """
+    columnNames = [c["name"] for c in inspection]
+    if spec.fields.names != set(columnNames):
+        raise RuntimeError(f"Table '{name}' exists but is defined differently in the database; "
+                           f"specification has columns {list(spec.fields.names)}, while the "
+                           f"table in the database has {columnNames}.")
+
+
+class TransactionInterruption(RuntimeError):
```

Comment:
I think exception classes should end with "Error". Same for `SynchronizationConflict` below.